xdsl/interpreter.py

```diff
@@ -265,9 +304,80 @@ def run(self, op: Operation):
         )
         self.set_values(zip(op.results, results))
 
+    def run_terminator(self, terminator: Operation):
+        """
+        Fetches the implemetation for the given terminator, passes it the Python values
+        associated with the SSA operands.
+        The Terminator implementation should return a Block | None, designating the target
+        successor if any, as well as values.
+        The returned values will be used as block arguments of the successor if any, or
+        returned as the parent region's results otherwise.
+        """
+        inputs = self.get_values(terminator.operands)
+        results = self._impls.run(self, terminator, inputs)
+        if len(terminator.successors) > 0:
+            successor = results[0]
+            results = results[1:]
+            self.interpreter_assert(
+                isinstance(successor, Block),
+                "Expected Terminator implementation to return selected successor.",
+            )
+            assert isinstance(successor, Block)
+            self.interpreter_assert(
+                len(successor.args) == len(results),
+                "Incorrect number of results for successor",
+            )
+            return successor, results
+        else:
+            parent = terminator.parent_op()
+            if parent is None:
+                raise InterpretationError(f"Terminator without parent or successor")
+            return results
+
+    def run_block(
+        self, block: Block, args: tuple[Any, ...]
+    ) -> tuple[Block | None, tuple[Any, ...]]:
+        """
+        Interpret a basic block, using `args` as the block argument values.
+        This will set those values in the interpreter's SSA context, and
+        execute the block operations in-order, finally interpretting its
+        Terminator and returning the target successor if any, as well as
+        the values to be passed to the successor or the parent region.
+        """
+        self.set_values(zip(block.args, args))
+        if block.ops.last is None or block.ops.first is None:
+            raise InterpretationError(f"Tried to interpret an empty block.")
+        current_op = block.ops.first
+        while current_op != block.ops.last:
```

Comment:
I think a for loop would work here
```suggestion
        for current_op in block.ops:
            if current_op is block.ops.last:
                break
            self.run_op(current_op)
```

Reply:
I really don't like the `for`/`break` style.. What I would love to see is an iterator exposed by Block, what do you think?
Something like `for current_op in block.non_terminator_ops`.